Source/GTMSessionFetcherService.m

```diff
@@ -224,6 +231,54 @@ - (GTMSessionFetcher *)fetcherWithURLString:(NSString *)requestURLString {
   return [self fetcherWithURL:url];
 }
 
+- (void)addHeaderDecorator:(id<GTMSessionFetcherHeaderDecorator>)decorator {
+  @synchronized(self) {
+    [_headerDecorators addPointer:(__bridge void *)decorator];
+  }
+}
+
+- (void)removeHeaderDecorator:(id<GTMSessionFetcherHeaderDecorator>)decorator {
+  @synchronized(self) {
+    NSUInteger i = 0;
+    for (id<GTMSessionFetcherHeaderDecorator> headerDecorator in _headerDecorators) {
+      if (headerDecorator == decorator) {
+        break;
+      }
+      ++i;
+    }
+    GTMSESSION_ASSERT_DEBUG(i < _headerDecorators.count, @"header decorator %@ must be passed to -addHeaderDecorator: before removing", decorator);
+    if (i < _headerDecorators.count) {
+      [_headerDecorators removePointerAtIndex:i];
+    }
+  }
+}
+
+- (void)decorateHeadersForRequest:(NSURLRequest *)request
+                          fetcher:(GTMSessionFetcher *)fetcher {
+  @synchronized(self) {
+    BOOL needsCompact = NO;
+    for (id<GTMSessionFetcherHeaderDecorator> headerDecorator in _headerDecorators) {
+      if (!headerDecorator) {
```

Comment:
I'm a bit leery calling into unknown external code while inside an `@synchronized` block. I wonder if it wouldn't be better to fetch `_headerDecorators.allObjects` into a local array and do the processing outside the synchronized block?

Reply:
Sure, done.